app/src/main/java/com/heartz/app/presentation/main/MainNavigator.kt

```diff
@@ -0,0 +1,69 @@
+package com.heartz.app.presentation.main
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.navigation.NavDestination
+import androidx.navigation.NavDestination.Companion.hasRoute
+import androidx.navigation.NavHostController
+import androidx.navigation.compose.currentBackStackEntryAsState
+import androidx.navigation.compose.rememberNavController
+import androidx.navigation.navOptions
+import com.heartz.app.presentation.graph.navigation.navigateToFigure
+import com.heartz.app.presentation.home.navigation.Home
+import com.heartz.app.presentation.home.navigation.navigateToHome
+import com.heartz.app.presentation.mypage.navigation.navigateToMypage
+import com.heartz.app.presentation.quest.navigation.navigateToQuest
+
+class MainNavigator(
+    val navController: NavHostController
+) {
+    private val currentDestination: NavDestination?
+        @Composable get() =
+            navController
+                .currentBackStackEntryAsState().value?.destination
+
+    val startDestination = Home
+
+    val currentTab: MainNavTab?
+        @Composable get() =
+            MainNavTab.find { tab ->
+                currentDestination?.hasRoute(tab::class) == true
+            }
+
+    fun navigate(tab: MainNavTab) {
+        val navOptions =
+            navOptions {
+                navController.currentDestination?.route?.let {
+                    popUpTo(it) {
+                        inclusive = true
+                        saveState = true
+                    }
+                }
+                launchSingleTop = true
+                restoreState = true
+            }
+        when (tab) {
+            MainNavTab.QUEST -> navController.navigateToQuest(navOptions)
+            MainNavTab.HOME -> navController.navigateToHome(navOptions)
+            MainNavTab.FIGURE -> navController.navigateToFigure(navOptions)
+            MainNavTab.MYPAGE -> navController.navigateToMypage(navOptions)
+        }
+    }
+
+    @Composable
+    fun showBottomBar() =
+        MainNavTab.contains {
+            currentDestination?.hasRoute(it::class) == true
+        }
+
+    fun navigateUp() {
+        navController.navigateUp()
```

Comment:
p1
성능 최적화를 위해서 이렇게 해줘야 한다고 알고 있습니다!
```suggestion
        navController::navigateUp
```

Reply:
@znayeonzn 이 부분은 함수 참조(`::navigateUp`)와 함수 호출(`navigateUp()`)의 차이에 대해 한 번 공부해보면 좋을 것 같아요! 😊

예를 들어 `Button`의 `onClick`처럼 **버튼이 눌렸을 때 동작해야 하는 상황**에서는 다음처럼 함수 참조를 사용해요

```kotlin
fun navigateUp(): Boolean {
    return navController.navigateUp()
}

Button(onClick = ::navigateUp) // 버튼을 누를 때 실행 됨.

Reply:
아하~ 종리드 천재. 떠먹어주셔서 감사합니다 🤩